README
This is the README file for the 'webget' package.

Webget is a micro web site for exporting audio off of the Rivendell system
at Radio Free Asia.


NEWS
This is the NEWS file for the 'webget' package.


htdocs/webget.js
// webget.js

function Id(id)
{
    return document.getElementById(id);
}


function GetMpegVersion(samprate)
{
    mpeg_ver=1.0;

    switch(samprate) {
    case 32000:
    case 44100:
    case 48000:
	mpeg_ver=1.0;
	break;

    case 16000:
    case 22050:
    case 24000:
	mpeg_ver=2.0;
	break;

    case 8000:
    case 11025:
    case 12000:
	mpeg_ver=2.5;
	break;
    }
    return mpeg_ver;
}


function SetLayerII()
{
    mpeg_ver=GetMpegVersion(parseInt(Id('samprate').value));
    if(mpeg_ver==1.0) {
	Id('bitrate').innerHTML=
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="48000" selected>48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n'+
	    '<option value="80000">80 kbit/sec</option>\n'+
	    '<option value="96000">96 kbit/sec</option>\n'+
	    '<option value="112000">112 kbit/sec</option>\n'+
	    '<option value="128000">128 kbit/sec</option>\n'+
	    '<option value="160000">160 kbit/sec</option>\n'+
	    '<option value="192000">192 kbit/sec</option>\n';
	    /* Bitrates higher than 192 kbit/sec are allowed only for
	     * stereo and dual channel modes.
	    '<option value="224000">224 kbit/sec</option>\n'+
	    '<option value="256000">256 kbit/sec</option>\n'+
	    '<option value="320000">320 kbit/sec</option>\n'+
	    '<option value="384000">384 kbit/sec</option>\n';
	    */
    }
    if(mpeg_ver==2.0) {
	Id('bitrate').innerHTML=
	    '<option value="8000">8 kbit/sec</option>\n'+
	    '<option value="16000" selected>16 kbit/sec</option>\n'+
	    '<option value="24000">24 kbit/sec</option>\n'+
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="40000">40 kbit/sec</option>\n'+
	    '<option value="48000">48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n'+
	    '<option value="80000">80 kbit/sec</option>\n'+
	    '<option value="96000">96 kbit/sec</option>\n'+
	    '<option value="112000">112 kbit/sec</option>\n'+
	    '<option value="128000">128 kbit/sec</option>\n'+
	    '<option value="144000">144 kbit/sec</option>\n'+
	    '<option value="160000">160 kbit/sec</option>\n';
    }
    if(mpeg_ver==2.5) {
	Id('bitrate').innerHTML=
	    '<option value="8000">8 kbit/sec</option>\n'+
	    '<option value="16000" selected>16 kbit/sec</option>\n'+
	    '<option value="24000">24 kbit/sec</option>\n'+
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="40000">40 kbit/sec</option>\n'+
	    '<option value="48000">48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n';
    }
}


function SetLayerIII()
{
    mpeg_ver=GetMpegVersion(parseInt(Id('samprate').value));
    if(mpeg_ver==1.0) {
	Id('bitrate').innerHTML=
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="40000" selected>40 kbit/sec</option>\n'+
	    '<option value="48000">48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n'+
	    '<option value="80000">80 kbit/sec</option>\n'+
	    '<option value="96000">96 kbit/sec</option>\n'+
	    '<option value="112000">112 kbit/sec</option>\n'+
	    '<option value="128000">128 kbit/sec</option>\n'+
	    '<option value="160000">160 kbit/sec</option>\n'+
	    '<option value="192000">192 kbit/sec</option>\n'+
	    '<option value="224000">224 kbit/sec</option>\n'+
	    '<option value="256000">256 kbit/sec</option>\n'+
	    '<option value="320000">320 kbit/sec</option>\n';
    }
    if(mpeg_ver==2.0) {
	Id('bitrate').innerHTML=
	    '<option value="8000">8 kbit/sec</option>\n'+
	    '<option value="16000" selected>16 kbit/sec</option>\n'+
	    '<option value="24000">24 kbit/sec</option>\n'+
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="40000">40 kbit/sec</option>\n'+
	    '<option value="48000">48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n'+
	    '<option value="80000">80 kbit/sec</option>\n'+
	    '<option value="96000">96 kbit/sec</option>\n'+
	    '<option value="112000">112 kbit/sec</option>\n'+
	    '<option value="128000">128 kbit/sec</option>\n'+
	    '<option value="144000">144 kbit/sec</option>\n'+
	    '<option value="160000">160 kbit/sec</option>\n';
    }
    if(mpeg_ver==2.5) {
	Id('bitrate').innerHTML=
	    '<option value="8000">8 kbit/sec</option>\n'+
	    '<option value="16000" selected>16 kbit/sec</option>\n'+
	    '<option value="24000">24 kbit/sec</option>\n'+
	    '<option value="32000">32 kbit/sec</option>\n'+
	    '<option value="40000">40 kbit/sec</option>\n'+
	    '<option value="48000">48 kbit/sec</option>\n'+
	    '<option value="56000">56 kbit/sec</option>\n'+
	    '<option value="64000">64 kbit/sec</option>\n';
    }
}


function samplerateChanged()
{
    if(Id('format').value=='MP2') {
	SetLayerII();
    }
    if(Id('format').value=='MP3') {
	SetLayerIII();
    }
}


function formatChanged()
{
    Id('bitrate').disabled=
	!((Id('format').value=="MP2")||
	  (Id('format').value=="MP3"));
    Id('quality').disabled=Id('format').value!="VORBIS";
    samplerateChanged();
}


htdocs/webget.py
#!/usr/bin/python

from __future__ import print_function

import os
import sys
import subprocess
import tempfile
import urllib

def eprint(*args,**kwargs):
    print(*args,file=sys.stderr,**kwargs)

title=''
samprate=48000
format='MP3'
bitrate=32000
quality=2

#
# Get call fields
#
f0=os.environ['QUERY_STRING'].split('&')
for field in f0:
    f1=field.split('=')
    if(f1[0]=='title'):
        title=urllib.unquote(f1[1])
    if(f1[0]=='samprate'):
        samprate=int(urllib.unquote(f1[1]))
    if(f1[0]=='format'):
        format=urllib.unquote(f1[1])
    if(f1[0]=='bitrate'):
        bitrate=int(urllib.unquote(f1[1]))
    if(f1[0]=='quality'):
        quality=int(urllib.unquote(f1[1]))
if(len(title)==0):
    print('Content-type: text/html')
    print()
    print('No title specified!')
    sys.exit(0)
title=title.replace('+',' ')
    
#
# Generate file export
#
tempdir=tempfile.mkdtemp()
try:
    filename=subprocess.check_output(('rdexport','--title='+title,
                                      '--metadata-pattern=%t',
                                      '--samplerate='+str(samprate),
                                      '--format='+format,
                                      '--bitrate='+str(bitrate),
                                      '--quality='+str(quality),
                                      tempdir))
except subprocess.CalledProcessError:
    os.rmdir(tempdir)
    print('Content-type: text/html')
    print('Status: 500')
    print()
    print('500 - unable to execute rdexport(1)')
    sys.exit(0)
if(len(filename)==0):
    os.rmdir(tempdir)
    print('Content-type: text/html')
    print('Status: 404')
    print()
    print('404 - no cart with that title');
    sys.exit(0)
filename=filename[0:len(filename)-1]
filepath=tempdir+'/'+filename
try:
    mimetype=subprocess.check_output(('file','--brief','--mime-type',filepath))
except subprocess.CalledProcessError:
    os.remove(filepath)
    os.rmdir(tempdir)
    print('Content-type: text/html')
    print('Status: 500')
    print()
    print('500 - unable to determine output mime-type')
    sys.exit(0)

#
# Render the output
#
try:
    f=open(filepath,'r')
except IOError, reason:
    print('Content-type: text/html')
    print()
    print('500 - unable to open exported file "'+filepath+'" ['+str(reason)+']')
    sys.exit(0)
    
print('Content-Disposition: attachment; filename=\"'+filename+'\"')
print('Content-type: '+mimetype)

bytes=f.read(1024)
while(len(bytes)>0):
    sys.stdout.write(bytes)
    bytes=f.read(1024)
f.close()
os.remove(filepath)
os.rmdir(tempdir)
